### ISSUE TITLE
Allow 'pre_knit' hooks to return new file to be used for knit?

### ISSUE THREAD

**Issue by kevinushey** [rstudio/rmarkdown#974]
Motivation: currently, HTML comments are not really properly handled by R Markdown or knitr. For example, with the following document:

    ---
    title: "Untitled"
    output: html_document
    ---

    

Even though the R chunk below is within an HTML comment, that chunk still gets executed, and thus included and processed by Pandoc.

When rendering such a document in `html_document` mode, those comment markers survive, and so the generated output is not shown in the document, giving the user the impression that the commented chunk was ignored.

However, with `html_notebook`, we're not so lucky -- the closing comment tag is actually HTML-escaped, and so the comment never actually functions as expected -- we end up seeing the chunk output in the rendered document.

Correct me if I am wrong, but since the current experience is as though HTML comments in R Markdown documents are supported, we need to do something extra to ensure that this works for HTML notebooks.

My initial thought would be that the `pre_knit` hook could return a file path that would then be passed on to `knitr::knit()`, and that the `pre_knit` hook could be in charge of e.g. replacing comments with a placeholder (and later restoring the placeholders in e.g. a `post_process` hook). However, this is not currently possible because of how the `pre_knit` hooks are overlaid (the union of return values seems to be returned?)

Is this something that could be made possible?

**Comment by jjallaire** [comment]
It looks like if we had the `pre_knit` function return a filename (as an alternate input) we could assign that to the `knit_input` variable and that would do the job. We should however inspect all uses of the `knit_input` variable to ensure that we don't need yet another variable that represents the file to be passed to knitr as distinct from `knit_input`.

As regards the union of `pre_knit` hooks, I'd just look for a filename and use the first or last one found (warning if more than one are found).

**Comment by MichaelChirico** [comment]
Approve! http://stackoverflow.com/questions/43784515/commenting-in-rmarkdown-knitr-to-prevent-r-evaluation?noredirect=1#comment74611248_43784515

**Comment by baptiste** [comment]
Tangential to this discussion, I recently discovered an alternative strategy to comment out sections of Rmd documents, which works for both text and chunks,

```
---
title: "Untitled"
output: html_notebook
---

---
# all this will be commented
# ```{r}
# print("Hello!")
# ```
---
```

I find it much nicer (with Ctrl+Shift+C) than the html comment tags, but it still doesn't prevent knitr from over-doing it with inline chunks.

**Comment by muschellij2** [comment]
So is it possible to do some preprocessing before knitting? I think this would be worthwhile as @jjallaire  alluded to above.  This would be helpful because otherwise you can't really preprocess the .Rmd/md file (such as stripping out sensitive information or something else tagged).  

Using `pre_knit` does something but not sure as to what from https://github.com/rstudio/rmarkdown/blob/master/R/render.R#L529 and the docs don't seem to indicate the behavior of of the output: https://github.com/rstudio/rmarkdown/blob/master/R/output_format.R#L26.   Hard to see how to do this also when defining or extending a format as in https://bookdown.org/yihui/rmarkdown/format-custom.html.  

I used to be able to do this by using `intermediates_generator` and then replacing the knitted file with one where `gsub` and other text processing removed some tags. 

I may be missing something, but I'm trying to help our users use the convenient `Knit` button, so have to make a output format, otherwise a wrapper around `render` could do what I want.

**Comment by jdblischak** [comment]
I agree this would be a nice feature to have. Currently my workflowr package uses a [hack](https://github.com/jdblischak/workflowr/blob/master/R/wflow_html.R#L315) to change the value of the internal variable `knit_input` in the call to `render()`. By setting the `knit_root_dir` and the output file, this enables it to knit a modified copy of the original file. I got this idea from [Romain's rmarkdowntown package](https://github.com/romainfrancois/rmarkdowntown/blob/deef97a5cd6f0592318ecc6e78c6edd7612eb449/R/html_document2.R#L12).

**Comment by yihui** [comment]
This feature still has not been implemented, but I just came up with another solution to comment out arbitrary content (including code chunks) in R Markdown, i.e., via the `comment` engine (which requires the dev version of **knitr**: `remotes::install_github('yihui/knitr')`), e.g.,

`````
---
title: "Untitled"
output: html_document
---

````{comment}

```{r}
print("Hello!")
```

````
`````

**Comment by kevinushey** [comment]
That is very clever 😄  I love it!

**Comment by yihui** [comment]
The `comment` engine has been included in knitr 1.37: https://github.com/yihui/knitr/releases/tag/v1.37 Not sure if this alternative solution helps with the original problem. I'm closing this issue for now, but we can re-open it if we really need to support HTML comments. Thanks!
